### PR TITLE
Adding/Deleting authorization and telemetry headers

### LINKED ISSUE #265: LargeFileUploadTask cors error
Description
Same issue as described at #220. I commented there as well, but the issue has been closed.
I'm having the exact same issue on the latest version 2.0.0. Only seems to happen for personal accounts trying to use LargeFileUploadTask for upload to onedrive. LargeFileUploadTask works fine when used with a business account uploading both to a sharepoint document library or onedrive.
Other requests like GET /drives/<drive-id>/items/children work perfectly.
Console Errors:
Access to fetch at 'https://api.onedrive.com/rup/24b5d554d8c0d12a/<longurlpartremoved>' from origin 'http://localhost:8081' has been blocked by CORS policy: Request header field sdkversion is not allowed by Access-Control-Allow-Headers in preflight response.
Steps to Reproduce
Using LargeFileUploadTask as described here
Expected behavior: Start file upload
Actual behavior: Fails with a Failed to fetch error message
Usage Information
SDK Version - 2.0.0

 Browser (Check, if using Browser version of SDK)

Browser Name - Chrome

Version - 79.0.3945.130 (64bit)
[AB#6002](https://microsoftgraph.visualstudio.com/0985d294-5762-4bc2-a565-161ef349ca3e/_workitems/edit/6002)

### PR DESCRIPTION
**Pull request by nikithauc**
Fixes #265 and #344 

Problem 1 - #265 reports a CORs error which was caused because the SDK added telemetry headers only for every request that goes out. The redirected non-Graph endpoint complained about the unexpected SDK_Version header and caused a CORS error.
Solution for Problem 1 - This PR includes logic to check in the `TelemetryHandler` if the URL is a Graph URL and if yes process to add the telemetry headers. If the URL is a non Graph URL and the telemetry headers are present, then the headers are deleted.

Problem 2 - #344 @peombwa Thank you for helping me reproduce this issue! 
On redirection, the `RedirectionHandler` is currently setting the AuthorizationHeader to undefined instead of completely dropping the header key. 
The existing code worked fine in a node environment but failed when using ExpressJS. 

Solution for Problem 2 - This PR includes change in the `RedirectionHandler` to delete Authorization header key if present.

### CODE REVIEW

File: src/GraphRequestUtil.ts [hunk 1]
```diff
@@ -58,3 +58,46 @@ export const serializeContent = (content: any): any => {
...
 	return content;
 };
+/**
+ * To hold list of the service root endpoints for Microsoft Graph and Graph Explorer for each national cloud.
+ */
+export const graphURLs = new Set<string>();
+
+// using an IIFE to populate the set object with the graph host names as Set(iterable:Object) is not supported in Internet Explorer
+(() => {
+	const urls = ["graph.microsoft.com", "graph.microsoft.us", "dod-graph.microsoft.us", "graph.microsoft.de", "microsoftgraph.chinacloudapi.cn"];
```
Comment (baywet) [inline src/GraphRequestUtil.ts:68]: Set initialization instead and core-js requirement for older browsers?

File: src/middleware/TelemetryHandler.ts [hunk 2]
```diff
@@ -66,21 +66,35 @@ export class TelemetryHandler implements Middleware {
...
+					if (clientRequestId === null) {
+						clientRequestId = generateUUID();
+						setRequestHeader(context.request, context.options, TelemetryHandler.CLIENT_REQUEST_ID_HEADER, clientRequestId);
+					}
+					let sdkVersionValue: string = `${TelemetryHandler.PRODUCT_NAME}/${PACKAGE_VERSION}`;
+					let options: TelemetryHandlerOptions;
+					if (context.middlewareControl instanceof MiddlewareControl) {
+						options = context.middlewareControl.getMiddlewareOptions(TelemetryHandlerOptions) as TelemetryHandlerOptions;
+					}
+					if (typeof options !== "undefined") {
```
Comment (baywet) [inline src/middleware/TelemetryHandler.ts:83]:
```suggestion
					if (options !== undefined) {
```

**Review by baywet — CHANGES_REQUESTED** [review]

File: spec/middleware/TelemetryHandler.ts [hunk 3]
```diff
@@ -100,6 +101,27 @@ describe("TelemetryHandler.ts", () => {
...
+					middlewareControl: new MiddlewareControl(),
+				};
+				dummyHTTPHandler.setResponses([okayResponse]);
+				await telemetryHandler.execute(context);
+				assert.equal(context.options.headers["client-request-id"], undefined);
+				assert.equal(context.options.headers["SdkVersion"], undefined);
+				assert.equal(context.options.headers["setFeatureUsage"], undefined);
+			} catch (error) {
+				throw error;
+			}
```
Comment (zengin) [inline spec/middleware/TelemetryHandler.ts:123]: Please remove the try catch block.

File: src/middleware/TelemetryHandler.ts [hunk 4]
```diff
@@ -66,21 +66,35 @@ export class TelemetryHandler implements Middleware {
...
+					}
+					appendRequestHeader(context.request, context.options, TelemetryHandler.SDK_VERSION_HEADER, sdkVersionValue);
+				} else {
+					// Remove telemetry headers if present during redirection.
+					if (context.options.headers[TelemetryHandler.CLIENT_REQUEST_ID_HEADER]) {
+						delete context.options.headers[TelemetryHandler.CLIENT_REQUEST_ID_HEADER];
+					}
+					if (context.options.headers[TelemetryHandler.SDK_VERSION_HEADER]) {
+						delete context.options.headers[TelemetryHandler.SDK_VERSION_HEADER];
+					}
```
Comment (zengin) [inline src/middleware/TelemetryHandler.ts:95]: I think you can call delete on the keys even if they don't exist. See [this example](https://www.typescriptlang.org/play?#code/PTAEHUFMBsGMHsC2lQBd5oBYoCoE8AHSAZVgCcBLA1UABWgEM8BzM+AVwDsATAGiwoBnUENANQAd0gAjQRVSQAUCEmYKsTKGYUAbpGF4OY0BoadYKdJMoL+gzAzIoz3UNEiPOofEVKVqAHSKymAAmkYI7NCuqGqcANag8ABmIjQUXrFOKBJMggBcISGgoAC0oACCoASMFmgY7p7ehCTkVOle4jUMdRLYTqCc8LEZzCZmoNJODPHFZZXVtZYYkAAeRJTInDQS8po+rf40gnjbDKv8LqD2jpbYoACqAEoAMsK7sUmxkGSCc+VVQQuaTwVb1UBrDYULY7PagbgUZLJH6QbYmJAECjuMigZEMVDsJzCFLNXxtajBDIKMjJHooWg-QTwLwAb0UJRKnAYyHyglQlE4zHZHIYzEgAH58px2IhpD9FABfADcwQQnD51UZzNA+ToWq8AF5QGyOYNuZBdQAiYhISAAOXNlt4wpKootoAAjAAGRUqxRqpnuALQeDMAAUGyZnAAlCruDBIApNb9mQBtS1uy0AXRVAfgQZD4cjzNjQA)
```suggestion
					delete context.options.headers[TelemetryHandler.CLIENT_REQUEST_ID_HEADER];
					delete context.options.headers[TelemetryHandler.SDK_VERSION_HEADER];
```

File: src/middleware/RedirectHandler.ts [hunk 5]
```diff
@@ -201,8 +201,8 @@ export class RedirectHandler implements Middleware {
 					delete context.options.body;
 				} else {
 					const redirectUrl: string = this.getLocationHeader(response);
-					if (!this.isRelativeURL(redirectUrl) && this.shouldDropAuthorizationHeader(response.url, redirectUrl)) {
-						setRequestHeader(context.request, context.options, RedirectHandler.AUTHORIZATION_HEADER, undefined);
+					if (!this.isRelativeURL(redirectUrl) && this.shouldDropAuthorizationHeader(response.url, redirectUrl) && context.options.headers[RedirectHandler.AUTHORIZATION_HEADER]) {
```
Comment (zengin) [inline src/middleware/RedirectHandler.ts:204]: see the example I gave below on delete.

**Review by zengin — CHANGES_REQUESTED** [review]

**Review by baywet — APPROVED** [review]
besides the comments @zengin already added, LGTM. Thanks for making the changes!

**Review by MIchaelMainer — COMMENTED** [review]
Oops, forgot to submit this yesterday

**Review by zengin — APPROVED** [review]

**Review by baywet — APPROVED** [review]

**Review by MIchaelMainer — APPROVED** [review]

**Pull request merged**